dbms/src/Storages/DeltaMerge/File/DMFile.cpp

```diff
@@ -625,18 +627,43 @@ void DMFile::finalizeForFolderMode(const FileProviderPtr & file_provider, const
     initializeIndices();
 }
 
+std::vector<String> DMFile::listLocal(const String & parent_path)
+{
+    Poco::File folder(parent_path);
+    std::vector<String> file_names;
+    if (folder.exists())
+    {
+        folder.list(file_names);
+    }
+    return file_names;
+}
+
+std::vector<String> DMFile::listS3(const String & parent_path)
+{
+    std::vector<String> filenames;
+    auto client = S3::ClientFactory::instance().sharedClient();
+    const auto & bucket = S3::ClientFactory::instance().bucket();
+    S3::listPrefix(*client, bucket, parent_path, [&filenames, &parent_path](const Aws::S3::Model::ListObjectsV2Result & result) {
```

Comment:
Seems here we need to use list with delimiter to get uniq dmfile id
Ref https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html